tests/_test_sso_localhost/test_sso_localhost.py

```diff
@@ -0,0 +1,390 @@
+import os
+import subprocess
+import time
+
+from urllib.parse import urljoin
+from bs4 import BeautifulSoup
+
+import requests
+import pytest
+
+os.environ['TEST_CASES_ENABLED'] = 'true'
+
+@pytest.fixture(scope='session')
+def client():
+    """
+    Fixture to set up and tear down the test client
+
+    This fixture sets up the test environment by performing the following steps:
+    1. Generates key and certificate files for the service provider (SP) and the
+    identity provider (IDP).
+    2. Starts the SP and downloads the metadata file from SP using curl.
+    3. Generates the IDP metadata for the SP.
+    4. Starts the IDP.
+    5. Configures the SP and creates necessary database tables.
+    6. Yields the test client object for making requests to the application.
+    7. Cleans up the environment by removing generated files, terminating the IDP process,
+       and resetting the environment variable.
+
+    The client fixture provides the test client object within a test context, allowing the tests
+    to interact with the application as a client.
+    """
+
+
+    # Configure flask service provider for testing
+
+    # Generate key and certificate for service provider (SP)
+    process_gen_crt_key_sp = (
+    'openssl req -x509 -newkey rsa:4096 -keyout mslib/auth_client_sp/test_key_sp.key '
+    '-out mslib/auth_client_sp/test_crt_sp.crt -sha256 -days 3650 -nodes '
+    '-subj "/C=XX/ST=State/L=City/O=Company/OU=CompanySection/CN=CommanName"'
+    )
+
+    subprocess.run(process_gen_crt_key_sp, shell=True, check=True)
+    # Generate key and certificate for identity provider (IDP)
+    process_gen_crt_key_idp = (
+    'openssl req -x509 -newkey rsa:4096 -keyout mslib/idp/test_key_idp.key '
+    '-out mslib/idp/test_crt_idp.crt -sha256 -days 3650 -nodes '
+    '-subj "/C=XX/ST=State/L=City/O=Company/OU=CompanySection/CN=CommonName"'
+    )
+
+    subprocess.run(process_gen_crt_key_idp, shell=True, check=True)
```

Comment:
a further step can be to use pyopenssl 
- https://anaconda.org/conda-forge/pyopenssl
- https://www.pyopenssl.org/en/latest/